File: problems/1046-last-stone-weight/go/solution.go
package solution

import "container/heap"

type maxHeap []int

func (h maxHeap) Len() int           { return len(h) }
func (h maxHeap) Less(i, j int) bool { return h[i] > h[j] }
func (h maxHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }
func (h *maxHeap) Push(x any)        { *h = append(*h, x.(int)) }
func (h *maxHeap) Pop() any {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[:n-1]
	return x
}

// Last Stone Weight — https://leetcode.com/problems/last-stone-weight/
func lastStoneWeight(stones []int) int {
	h := &maxHeap{}
	*h = append(*h, stones...)
	heap.Init(h)
	for h.Len() > 1 {
		a := heap.Pop(h).(int)
		b := heap.Pop(h).(int)
		if a != b {
			heap.Push(h, a-b)
		}
	}
	if h.Len() == 0 {
		return 0
	}
	return (*h)[0]
}

File: problems/1046-last-stone-weight/typescript/solution.ts
class MaxHeap {
  private data: number[] = [];

  size(): number {
    return this.data.length;
  }

  push(val: number): void {
    this.data.push(val);
    let i = this.data.length - 1;
    while (i > 0) {
      const parent = (i - 1) >> 1;
      if (this.data[parent] >= this.data[i]) break;
      [this.data[parent], this.data[i]] = [this.data[i], this.data[parent]];
      i = parent;
    }
  }

  pop(): number {
    const top = this.data[0];
    const last = this.data.pop();
    if (last !== undefined && this.data.length > 0) {
      this.data[0] = last;
      let i = 0;
      const n = this.data.length;
      while (true) {
        let largest = i;
        const l = 2 * i + 1;
        const r = 2 * i + 2;
        if (l < n && this.data[l] > this.data[largest]) largest = l;
        if (r < n && this.data[r] > this.data[largest]) largest = r;
        if (largest === i) break;
        [this.data[largest], this.data[i]] = [this.data[i], this.data[largest]];
        i = largest;
      }
    }
    return top;
  }
}

// Last Stone Weight — https://leetcode.com/problems/last-stone-weight/
export function lastStoneWeight(stones: number[]): number {
  const heap = new MaxHeap();
  for (const s of stones) heap.push(s);
  while (heap.size() > 1) {
    const a = heap.pop();
    const b = heap.pop();
    if (a !== b) heap.push(a - b);
  }
  return heap.size() === 0 ? 0 : heap.pop();
}
